servo/connectors/newrelic.py

```diff
@@ -0,0 +1,340 @@
+    
+    @validator("base_url", allow_reuse=True)
+    @classmethod
+    def rstrip_base_url(cls, base_url):
+        return base_url.rstrip("/")
+
+    @property
+    def api_url(self) -> str:
+        return f"{self.base_url}{API_PATH}"
+
+class NewrelicRequest(BaseModel):
+    base_url: AnyHttpUrl
+    fetches: Dict[str, Set[str]]
+    start: datetime
+    end: datetime
+    step: Duration
+
+    @property
+    def params(self) -> Dict[str, str]:
+        return { 
+            'names[]': '+'.join(self.fetches.keys()), # TODO can these be lists and httpx will extrapolate?
+            'values[]': '+'.join(reduce(lambda x, y: x+y, self.fetches.values())),
+            'from': self.start.isoformat(),
+            'to': self.end.isoformat(),
+            'period': self.step.total_seconds(),
+            'summarize': False,
+            'raw': True,
+        }
+
+class NewrelicChecks():
+    pass # TODO
+
+@metadata(
+    description="Newrelic Connector for Opsani",
+    version="0.0.1",
+    homepage="https://github.com/opsani/newrelic-connector",
+    license=License.APACHE2,
+    maturity=Maturity.EXPERIMENTAL,
+)
+class NewrelicConnector(BaseConnector):
+    """NewrelicConnector objects enable servo assemblies to capture
+    measurements from the [Newrelic](https://newrelic.com/) metrics server.
+    """
+    config: NewrelicConfiguration
+
+    # TODO
+    # @on_event()
+    # async def check(self,
+    #     filter_: Optional[Filter] = None, 
+    #     halt_on: HaltOnFailed = HaltOnFailed.requirement
+    # ) -> List[Check]:
+    #     """Checks that the configuration is valid and the connector can capture        
+    #     measurements from Newrelic.
+
+    #     Checks are implemented in the NewrelicChecks class.
+
+    #     Args:
+    #         filter_ (Optional[Filter], optional): A filter for limiting the
+    #             checks that are run. Defaults to None.
+    #         halt_on (HaltOnFailed, optional): When to halt running checks.
+    #             Defaults to HaltOnFailed.requirement.
+
+    #     Returns:
+    #         List[Check]: A list of check objects that report the outcomes of the            
+    #             checks that were run.
+    #     """        
+    #     return await NewrelicChecks.run(self.config, filter_, halt_on=halt_on)
+
+    @on_event()
+    def describe(self) -> Description:
+        """Describes the current state of Metrics measured by querying Newrelic.
+
+        Returns:
+            Description: An object describing the current state of metrics
+                queried from Newrelic.
+        """
+        return Description(metrics=self.config.metrics)
+
+    @property
+    @on_event()
+    def metrics(self) -> List[Metric]:
+        """Returns the list of Metrics measured through Newrelic queries.
+
+        Returns:
+            List[Metric]: The list of metrics to be queried.
+        """
+        return self.config.metrics
+
+    @on_event()
+    def instances(self) -> List[str]:
```

Comment:
Since that endpoint is freeform and has been replaced in `oco-e`, this isn't the right place for this functionality. I'd expect this to live in another connector like Harness or whatever that is responsible for providing the instances and New Relic connectr would just call out to it rather than trying to provide an implementation itself.

Reply:
It API requests to newrelic right here: https://github.com/opsani/servox/blob/bc8d66cd934ddac09978541bc8d8471a473bf788/servo/connectors/newrelic.py#L224

If it weren't for that, I would have put it in a standalone connector

Reply:
All of this business is potentially quite brittle, coupled to EC2 implementation details, and requires some other out of band actor to populate the data in the `/config` endpoint: https://github.com/opsani/servox/blob/bc8d66cd934ddac09978541bc8d8471a473bf788/servo/connectors/newrelic.py#L210-L222

Rather than putting in an explicit switch on the configuration (though now that I think about it allowing a hard-coded `instances` list may be useful), what I was thinking is using the presence of one or more `instances` event handlers to override the default behavior of just grabbing the full instances list. This would let you evict the `/config` and EC2 code smells and switch the behavior based on if there are any results returned from a call to `dispatch_event("instances")`.

While I generally agree with your observation about the relatively limited utility of a default implementation that returns all the instances in today's state of affairs with Intuit and Ancestry, as we take Opsani Dev to market and have more middle-market customers my thinking is that we will start to see more engagements where the customer only has one main application and all the monitoring infrastructure is dedicated to it.

My New Relic API knowledge is a bit rusty, but I also wonder if we can eliminate the need for custom `instances` handlers in 80% of the use cases by supporting optional filtering predicates in the config class. There is an `application_name` attribute in the response payload (see https://docs.newrelic.com/docs/apis/rest-api-v2/get-started/list-application-id-host-id-instance-id#locating_instance_id and https://docs.newrelic.com/docs/apis/rest-api-v2/get-started/list-application-id-host-id-instance-id#locating_instance_id) -- if we supported an exact string matcher and regex pattern, shouldn't that cover most sane implementations you can imagine?

Reply:
@derezed88 Could we get your thoughts and commentary on the last paragraph of Blake's latest response?

Reply:
I had a long conversation with Ben this afternoon trying to get my head around the specific challenges/requirements for this instance business in the context of Harness.

I have recommended that we explore the idea of using New Relic's support of multiple application names which we can set via the `NEW_RELIC_APP_NAME` environment variable in the form of  `app-name1;app-name2`.

If there are not any blockers that I can't see at the moment (@benburdick is testing these ideas out), I think we can factor out the special snowflakieness of this implementation.

I have also been looking into New Relic labels/tags as another supporting mechanism for some of the stuff we are doing but this probably starts to intersect more with the currently theoretical Harness servox connector